src/installer/protocols.py

```diff
@@ -0,0 +1,120 @@
+import abc
+from enum import Enum
+
+from installer._compat import with_metaclass
+from installer._compat.typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from io import BufferedReader
+    from typing import List, Sequence, Tuple
+
+    from installer._compat.typing import Text, FSPath
+
+
+Scheme = Enum("Scheme", ["purelib", "platlib", "headers", "scripts", "data"])
+
+
+@with_metaclass(abc.ABCMeta)
+class WheelSource(object):
+    """Represents an installable wheel.
+    """
+
+    def __init__(self, distribution, version):
+        # type: (Text, Text) -> None
+        super(WheelSource, self).__init__()
+        self.distribution = distribution
+        self.version = version
+
+    @property
+    def dist_info_dir(self):
+        """Name of the dist-info directory.
+        """
+        return u"{}-{}.dist-info".format(self.distribution, self.version)
+
+    @property
+    def data_dir(self):
+        """Name of the data directory.
+        """
+        return u"{}-{}.data".format(self.distribution, self.version)
+
+    @abc.abstractproperty
+    def dist_info_filenames(self):
+        # type: () -> List[FSPath]
+        """Get names of all files in the dist-info directory.
+
+        Sample usage/behavior::
+
+            >>> wheel_source.dist_info_filenames
+            ["INSTALLER", "METADATA", "WHEEL"]
+        """
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def open_dist_info(self, filename):
+        # type: (FSPath) -> BufferedReader
+        """Get readable stream of data, from ``filename`` in the dist-info directory.
+
+        Sample usage/behavior::
+
+            >>> f = wheel_source.open_dist_info("INSTALLER")
+            >>> f.read()
+            'pip'
+        """
+        raise NotImplementedError()
+
+    # All files in the wheel
+    @abc.abstractmethod
+    def iter_files(self):
```

Comment:
I will say I dislike `iter_*()` API names as it doesn't communicate what the method actually does, just what you are expected to do with it or what the turn type is depending if you read that as shorthand as `iterate` or `iterator`, respectively. I personally would be fine with `files()` and simply document it returns an iterator.

Reply:
Personally I like having `iter_` since the fact that an iterator is returned is the second most important part of what the function does (the first is it iterators through files). Python does not offer any way to guard against iterator reuse, and the only thing to prevent accidental usage is to stick something into the name. Type-checking does not help here because it would happily accept usages like

```python
def install_files(files: Iterable[File]):
   ...

files = source.files()
logger.debug('Processing %s', [path for path, _ in files])
install_files(files)
```

I won’t speak for others, but I can’t tell whether this code is correct or not in a quick code review. This works if `files()` returns a list, but not if it returns an iterator. Maybe I should call `files()` twice? But that wouldn make things worse if the function implies significant performance impact. I can go through this thought process every time I see this, and potentially dig into the documentation or source code. Or I can easily avoid all the mental work by simply adding `iter_` to the function name, with a slight aesthetics cost. The trade-off is worth it IMO.

Reply:
I'm gonna have to change this for timestamp + record-keeping reasons.

For now, I'm think gonna call this `get_contents`, returning a `Tuple[Record, BufferedReader, Optional[stat_result]]`. I'm a bit concerned with using `Record` here directly, since @agronholm has noted that he wants to be able to make `wheel`'s `WheelFile` capable of being a `WheelSource`.

Reply:
So are we going to restore the modification time and file attributes present in the archive when the wheel is being installed? Sadly the wheel PEP does not say anything about this which I feel is a significant problem.

Reply:
Looking at [pip's logic](https://github.com/pypa/pip/blob/acab2ee54ed25edd44edbdf793a4b1d173bfbcb1/src/pip/_internal/operations/install/wheel.py#L665), we unpack the zip with the following guarentees:

> All files are written based on system defaults and umask (i.e. permissions are not preserved), except that regular file members with any execute permissions (user, group, or world) have "chmod +x" applied after being written. Note that for windows, any execute changes using os.chmod are no-ops per the python docs.

And then install from the unpacked wheel with, among other things:

```py
                # Copy over the metadata for the file, currently this only
                # includes the atime and mtime.
                st = os.stat(srcfile)
                if hasattr(os, "utime"):
                    os.utime(destfile, (st.st_atime, st.st_mtime))
```

I'm very confused right now about what the potential details we'd need, since IIUC, pip's not preserving the details during unpacking, but preserving them when moving the files (!).

Reply:
I have so many questions:

- Can we have empty folders in a wheel? If yes, how should they get installed?
- What metadata about a file in a wheel, do we actually care about?
	- Executable bits?
	- What properties from `os.stat` / `ZipInfo`?
- Would people be OK with files being written with timestamp=0? (this would be a concrete Destination implementation thing, so I'm not too worried)

Reply:
The answers to these questions should be added to the wheel spec, by the way 🙂

Reply:
> Can we have empty folders in a wheel? If yes, how should they get installed?

No. [There is a Dicourse thread proposing to amend this.](https://discuss.python.org/t/refine-wheel-spec-to-designate-directory-entries/1667)

> Would people be OK with files being written with `timestamp=0`? 

This would cause pain when integrating wheels into timestamp-based incremental build systems (e.g. GNU Make). I think setting timestamp to 0 would be worse than what pip currently does (which sets the timestamp to whenever the wheel is installed, I believe?)

Reply:
> > Can we have empty folders in a wheel? If yes, how should they get installed?
>
> No

Well, I'm gonna implement status quo, so not implementing any support for this sounds good. :)

> timestamp-based incremental build systems

To be clear, the situation would be that WheelSource objects wouldn't provide timestamp information, so Destination objects would set the timestamp based on whatever logic they want - I imagine pip would preserve its current approach and keep the time dependent on unpack time.

The question really should've been, do we care about preserving timestamps from a WheelSource when installing from it? I think the answer is no, so I'm going to move forward with that assumption.

Reply:
To me it seems correct to leave the timestamp in newly created files alone. If pip indeed does this already, then that opens the door for wheel to use the zip default timestamp for all files (instead of copying the timestamp from the file system). This too should be clarified in the new PEP which @uranusjr is apparently writing now.